File: src/org/jetbrains/plugins/scala/facet/ScalaFacetType.java
/*
 * Copyright 2000-2008 JetBrains s.r.o.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.plugins.scala.facet;

import com.intellij.facet.FacetType;
import com.intellij.facet.Facet;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleType;
import com.intellij.openapi.module.StdModuleTypes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.plugins.scala.ScalaIcons;

import javax.swing.*;

/**
 * @author ilyas
 */
public class ScalaFacetType extends FacetType<ScalaFacet, ScalaFacetConfiguration> {

  public static final ScalaFacetType INSTANCE = new ScalaFacetType();

  public ScalaFacetConfiguration createDefaultConfiguration() {
    return new ScalaFacetConfiguration();
  }

  private ScalaFacetType() {
    super(ScalaFacet.FACET_TYPE_ID, "Scala", "Scala");
  }

  public ScalaFacet createFacet(@NotNull Module module, String name, @NotNull ScalaFacetConfiguration configuration, @Nullable Facet underlyingFacet) {
    return new ScalaFacet(this, module, name, configuration, underlyingFacet);
  }

  public Icon getIcon() {
    return ScalaIcons.FILE_TYPE_LOGO;
  }

  public boolean isOnlyOneFacetAllowed() {
    return true;
  }

  public boolean isSuitableModuleType(ModuleType moduleType) {
    return moduleType == StdModuleTypes.JAVA || "PLUGIN_MODULE".equals(moduleType.getId());
  }
}

File: src/org/jetbrains/plugins/scala/config/ScalaFacetType.scala
package org.jetbrains.plugins.scala.config

import com.intellij.facet.{Facet, FacetConfiguration, FacetType}
import com.intellij.ide.util.projectWizard.ModuleBuilder
import com.intellij.openapi.module.{JavaModuleType, Module, ModuleType}
import org.jetbrains.plugins.scala.icons.Icons

/**
 * Pavel.Fatin, 26.07.2010
 */

class ScalaFacetType extends FacetType[ScalaFacet, ScalaFacetConfiguration](ScalaFacet.Id, "scala", "Scala") {
  override def getIcon = Icons.SCALA_SMALL_LOGO
  
  def isSuitableModuleType(moduleType: ModuleType[_ <: ModuleBuilder]) = 
    moduleType.isInstanceOf[JavaModuleType] || moduleType.getId == "PLUGIN_MODULE" || moduleType.isInstanceOf[ScalaFacetAvailabilityMarker]
  
  def createDefaultConfiguration = {
    new ScalaFacetConfiguration
  }

  def createFacet(module: Module, name: String, configuration: ScalaFacetConfiguration,
                  underlyingFacet: Facet[_ <: FacetConfiguration]) = {
    new ScalaFacet(module, name, configuration, underlyingFacet)
  }

  // workaround for FacetEditorFacadeImpl.addFacetNode(Facet) that uses object equality to compare facet types
  override def equals(obj: Any) = obj match {
    case facetType: ScalaFacetType => getStringId == facetType.getStringId
    case _ => false
  }
}